pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentMessageExpiryMonitor.java

```diff
@@ -100,6 +108,33 @@ public boolean expireMessages(int messageTTLInSeconds) {
         }
     }
 
+    private long getEntryTimestamp(Entry entry,
+                                  Map<Long, MLDataFormats.ManagedLedgerInfo.LedgerInfo> ledgerIdToInfoCache,
+                                  ManagedCursor cursor) throws IOException {
+        long entryTimestamp = Commands.getEntryTimestamp(entry.getDataBuffer());
+        ManagedLedger managedLedger = cursor.getManagedLedger();
+        ManagedLedgerConfig config = managedLedger.getConfig();
+        MLDataFormats.ManagedLedgerInfo.LedgerInfo ledgerInfo = ledgerIdToInfoCache.computeIfAbsent(
+                entry.getLedgerId(), (ledgerId) -> {
+                    try {
+                        return managedLedger.getLedgerInfo(ledgerId)
+                                .get(config.getMetadataOperationsTimeoutSeconds(), TimeUnit.SECONDS);
+                    } catch (Exception e) {
+                        return null;
+                    }
+                });
+        if (ledgerInfo == null) {
+            return entryTimestamp;
+        }
+        long maxAvailableTimestamp = ledgerInfo.getTimestamp() + config.getMaximumRolloverTimeMs();
```

Comment:
The `ledgerInfo.getTimestamp()` is the ledger closure time. Do we need to consider the maximum rollover time?